File: VB/E5030/Reports/ProductsByCategoriesReport.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports DevExpress.XtraReports.UI
Imports System.Web

Namespace E5030.Reports
	Partial Public Class ProductsByCategoriesReport
		Inherits DevExpress.XtraReports.UI.XtraReport
		Private Const ExpandedCategories_SessionKey As String = "ProductsByCategoriesReport_ExpandedCategories"

		Private _expandedCategories As HashSet(Of Integer) = Nothing

		Private ReadOnly Property ExpandedCategories() As HashSet(Of Integer)
			Get
				If _expandedCategories Is Nothing Then
					_expandedCategories = CType(HttpContext.Current.Session(ExpandedCategories_SessionKey), HashSet(Of Integer))
					If _expandedCategories Is Nothing Then
						_expandedCategories = New HashSet(Of Integer)()
						HttpContext.Current.Session(ExpandedCategories_SessionKey) = _expandedCategories
					End If
				End If
				Return _expandedCategories
			End Get
		End Property

		Public Sub New()
			InitializeComponent()
		End Sub

		Public Sub CategoryExpandClicked(ByVal categoryID As Integer)
			If ExpandedCategories.Contains(categoryID) Then
				ExpandedCategories.Remove(categoryID)
			Else
				ExpandedCategories.Add(categoryID)
			End If
		End Sub

		Private Sub lbExpand_HtmlItemCreated(ByVal sender As Object, ByVal e As HtmlEventArgs) Handles lbExpand.HtmlItemCreated
			e.ContentCell.Attributes.Add("onclick", String.Format("parent.ProductsByCategoriesReport_ExpandClick('{0}')", e.Brick.Value))
			e.ContentCell.Style.Add("cursor", "pointer")
		End Sub

		Private Sub DetailReport_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles DetailReport.BeforePrint
			Dim categoryID As Integer = GetCurrentColumnValue(Of Integer)("CategoryID")
			e.Cancel = Not ExpandedCategories.Contains(categoryID)
		End Sub

		Private Sub lbExpand_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles lbExpand.BeforePrint
			Dim label As XRLabel = TryCast(sender, XRLabel)
			Dim categoryID As Integer = GetCurrentColumnValue(Of Integer)("CategoryID")
			label.Text = If(ExpandedCategories.Contains(categoryID), "-", "+")
		End Sub

	End Class
End Namespace

File: CS/E5030/Reports/ProductsByCategoriesReport.cs
using System;
using System.Drawing;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using DevExpress.XtraReports.UI;
using System.Web;

namespace E5030.Reports {
    public partial class ProductsByCategoriesReport : DevExpress.XtraReports.UI.XtraReport {
        const string ExpandedCategories_SessionKey = "ProductsByCategoriesReport_ExpandedCategories";

        HashSet<int> _expandedCategories = null;

        HashSet<int> ExpandedCategories {
            get {
                if (_expandedCategories == null) {
                    _expandedCategories = (HashSet<int>) HttpContext.Current.Session[ExpandedCategories_SessionKey];
                    if (_expandedCategories == null) {
                        _expandedCategories = new HashSet<int>();
                        HttpContext.Current.Session[ExpandedCategories_SessionKey] = _expandedCategories;
                    }
                }
                return _expandedCategories;
            }
        }

        public ProductsByCategoriesReport() {
            InitializeComponent();
        }

        public void CategoryExpandClicked(int categoryID) {
            if (ExpandedCategories.Contains(categoryID)) {
                ExpandedCategories.Remove(categoryID);
            }
            else {
                ExpandedCategories.Add(categoryID);
            }
        }

        private void lbExpand_HtmlItemCreated(object sender, HtmlEventArgs e) {
            e.ContentCell.Attributes.Add("onclick", String.Format("parent.ProductsByCategoriesReport_ExpandClick('{0}')", e.Brick.Value));
            e.ContentCell.Style.Add("cursor", "pointer");
        }

        private void DetailReport_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            int categoryID = GetCurrentColumnValue<int>("CategoryID");
            e.Cancel = !ExpandedCategories.Contains(categoryID);
        }

        private void lbExpand_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            XRLabel label = sender as XRLabel;
            int categoryID = GetCurrentColumnValue<int>("CategoryID");
            label.Text = ExpandedCategories.Contains(categoryID) ? "-" : "+";
        }

    }
}
